### ISSUE TITLE
Kiwi's UA

### ISSUE THREAD

**Issue by IcedWatermelonJuice** [kiwibrowser/src.next#428]
I want to distinguish between chrome and kiwi, but Kiwi's UA is so similar to chrome that it is difficult to distinguish.

**Comment by KaKi87** [comment]
Why would anyone want that, unless you'd want to maliciously and discriminatorily inflict degraded user experience to Kiwi Browser users ?

**Comment by danny0838** [comment]
It is obvious that the essential of the user agent is to provide a meaningful expression of the user agent that is distinguishable from a different one. It is also very helpful for the developer to make the code more compatible to different browsers, which may have very different behaviors in some aspects that cannot be detected by any available API.

As a matter of fact, most Chomium-based browsers, such as Opera, Vilvadi, Brave, and new Edge, have a distinguishable UA from Chromium and Google Chrome.  I don't think Kiwi is any different from them.

Although it is possible that some bad sites use the UA in the wrong malicious or discriminatory way, it should be relatively rare as breaking user experience is generally doing nothing good for the web site owner. There are also user agent switcher extensions available for customizing the UA for such special cases. IMHO, Faking the UA natively is like throwing away the apple because of the core.

**Comment by KaKi87** [comment]
> It is also very helpful for the developer to make the code more compatible to different browsers

I would agree if we were talking about distinguishing browsers that use different rendering engines, like Chrome (Blink, formerly Webkit) vs Firefox (Quantum, formerly Gecko) vs Safari (Webkit).

But in this case, Kiwi uses Chrome, so everything that works on Chrome, must work on Kiwi, or be reported here. Therefore, there is no point in differenciating Kiwi from Chrome, unless you want to implement a dirty workaround.

Moreover, Kiwi Browser is the only open source Chromium-based Android browser that allows unlimited access to Chrome extensions, therefore hostile website developers would be more interested in discriminating it than others, especially to counteract content blockers.

**Comment by danny0838** [comment]
There may be difference s long as Kiwi is a **different** browser, even if it uses that same engine.

For example, can the bug I just reported in #427 be fixed? Are the developers willing to fix it? How long does it take to be fixed? If it can't be fixed in no time for any reason, there would be a need for the web site developer to use the user agent to deal with such issue for better browser compatibility.

It might be possible that a feature detection can work for this, but not every bug can be dealt with feature detection. OTOH, if a feature detection is possible for such thing, a malicious site can also use it do the same thing for a discriminatory purpose. After all, faking user agent won't actually work as long as there is any detectable feature difference among the different browsers.

**Comment by danny0838** [comment]
@gnadelwartz I agree that UA detection is generally the worst thing. Unfortunately, it is the only thing the developer can do in many cases, and many of them is what I have actually done in my projects.

About the point that "Google plans to remove UA": The User Agent Client Hints project Google has planned is to make a more accurate version of UA and make developers detect the user agent more easily, instead of making it impossible. This clearly is a dis-support instead of a support for the argument "UA is useless".

**Comment by gnadelwartz** [comment]
let me phrase it this way: classic UA-string us exactly this,: a string of unstructured content, you can try to extract information but you will fail for whatever reason. relying on this random string is at least ugly.

UA hints are http response headers containing information in a well defined format, even it may a bad idea to use this for bug or feature detection, it is not based on parsing a random string.

as always opinions may vary

**Comment by danny0838** [comment]
@gnadelwartz If I understand it correctly, this topic is about whether to provide a true user agent conventionally or to mimic/fake Chrome user agent, regardless of using the traditional UA or the new UA hints proposed by Google. What is your opinion about this?

**Comment by gnadelwartz** [comment]
I'm uncertain, as this may break random sites/parsing script's while the advantage is only for a small subset who want to optimize for kiwi.

e.g. kiwi has a special UA implemented for Facebook because Facebook parse UA instead of screen resolution.
the downside iis that now Facebook Always redirect to mobile even I use kiwi on tablet.

in my opinion kiwi is 100% chromium with desktop features enabled, therefore it's valid to use chromium UA. non chromium compatible HTML/CSS/JS behvior as in #425  is a bug and must be fixed.

**Comment by MTHS1901** [comment]
I did some tests comparing differences between "window" elements in Chrome PC/Android browser and Kiwi Browser, and for some reason in Kiwi Browser navigator.userAgentData.brands[2] is set to "undefined" so you could use the code below to set whether it is Kiwi browser or not.

!!chrome && navigator.userAgentData.brands[2] == undefined

!!chrome && !!chrome.runtime no longer works, after 106 update.
